### ISSUE TITLE
`Pattern` or `Regex` type

### ISSUE THREAD

**Issue by samuelcolvin** [annotated-types/annotated-types#57]
It would be nice to be able to enforce a regex pattern on a string.

In Pydantic this is already possible, but requires the slightly ugly syntax of

```py
from typing import Annotated
from pydantic import Field

foo: Annotated[str, Field(pattern='...')]
```

It would be nice it annotated-types provided a `Pattern()` type.

Perhaps due to the risk of confusion with `typing.Pattern`, we should use `RegexPattern`?

**Comment by Zac-HD** [comment]
See https://github.com/annotated-types/annotated-types/issues/9 - same arguments apply IMO; it's obviously useful downstream but too overloaded for me to be happy putting it in this package.

**Comment by samuelcolvin** [comment]
Hi Zac, hope you're well. What timezone are you in? 👀 

Sorry I missed that.

Personally I would argue that "practicality beats purity". Since downstream libraries have to deal with this nuances, not having a regex/pattern type in annotated-types doesn't make the situation worse.

In Pydantic as of 2.5 we'll have to flavours of regex - Rust and Python, we might even add more in the future.

I would suggest we add back `RegexPattern` but add another argument `flavor` which defaults to `'python'`.

If no one else agrees, happy to close this.

**Comment by adriangb** [comment]
I also agree that practically beats purity in this case. But I also do think there's some nuance around what that means in this case (`flavor` argument? not specifying it at all for now and leaving it up to implementers to decide what to do? always saying it's Python semantics and implements should either comply or raise an error if the pattern is not valid for their regex engine or is ambiguous)? I don't necessarily feel qualified enough about regexes to answer these questions.

**Comment by Zac-HD** [comment]
Pacific time at the moment, but heading over to the US east coast pretty regularly at the moment and I'll spend xmas + January in Australia... it's a timezone mess, really.  Going well despite all that though!

On the topic at hand, why not `from pydantic import RegexPattern`?  It seems like that would solve the immediate challenge, without opening up all of the difficult questions of what should happen in the flavor x library different cases.  You could even implement that in terms of `Predicate(re.compile("abc").search)`, which is unambiguous and can be supported by Hypothesis and Crosshair too.

**Comment by iron3oxide** [comment]
Is there any update on this? My use case: I'm currently building a sort of microframework that aims to have Pydantic (or rather SQLModel) models as the single source of truth for both the database and UI. As you can imagine, this requires me to annotate my fields a bit and also read the pydantic metadata (e.g. to use `title` as the label for a UI field). When it comes to validation though, passing `Field(..., gt=3, ...)` to `Annotated`, a `Gt(3)` object is produced in the pydantic metadata, while passing `Field(..., regex=r"abc", ...)` does not produce anything and in fact gets completely swallowed by the Field function as it passes it down to Rust. This in turn means I cannot easily use the regex in my model annotation as an input filter in the UI for the same field. Of course I could use an `AfterValidator` containing a callable class containing the regex, but that seems a bit inelegant to me compared to having a dedicated type.

**Comment by Zac-HD** [comment]
No update, I think we're basically blocked on a design proposal which addresses the challenges listed in #9.
